src/metrics/encoding/protobuf/reset_test.go

```diff
@@ -178,6 +178,83 @@ func TestReuseMetricWithMetadatasProtoOnlyCounter(t *testing.T) {
 	require.True(t, cap(input.CounterWithMetadatas.Metadatas.Metadatas) > 0)
 }
 
+func TestReuseMetricForEncoding(t *testing.T) {
+	input := metricpb.MetricWithMetadatas{
+		Type: metricpb.MetricWithMetadatas_COUNTER_WITH_METADATAS,
+		CounterWithMetadatas: &metricpb.CounterWithMetadatas{
+			Counter: metricpb.Counter{
+				Id: []byte("testCounter"),
+			},
+			Metadatas: metricpb.StagedMetadatas{
+				Metadatas: []metricpb.StagedMetadata{
+					{
+						Metadata: metricpb.Metadata{
+							Pipelines: []metricpb.PipelineMetadata{
+								{
+									ResendEnabled: true,
+									StoragePolicies: []policypb.StoragePolicy{
+										{
+											Resolution: policypb.Resolution{
+												WindowSize: 1000,
+											},
+											Retention: policypb.Retention{
+												Period: 1000,
+											},
+										},
+									},
+									Pipeline: pipelinepb.AppliedPipeline{
+										Ops: []pipelinepb.AppliedPipelineOp{
+											{
+												Rollup: pipelinepb.AppliedRollupOp{
+													Id: []byte("foobar"),
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+					{
+						CutoverNanos: 5678,
+					},
+				},
+			},
+		},
+	}
+	b, err := input.Marshal()
+	require.NoError(t, err)
+
+	var pb metricpb.MetricWithMetadatas
+	require.NoError(t, pb.Unmarshal(b))
+	require.Equal(t, input, pb)
+
+	// verify all slices are empty with a non-zero capacity.
+	ReuseMetricWithMetadatasProto(&pb)
+	metas := pb.CounterWithMetadatas.Metadatas.Metadatas
+	require.True(t, cap(metas) > 0)
```

Comment:
Write a helper for this?

Reply:
In case it's not clear, I meant the validations.

Reply:
what would the helper be? we'd need a separate helper for each slice type since go doesn't have generics.

Reply:
thanks golang